### DISCUSSION TITLE
Using drift from multiple isolates and databases with watch

### DISCUSSION THREAD

**Discussion by cn-cn-cn** [simolus3/drift#2815] [Q&A] [answered]
Hello,

I'm using the Drift feature of [running in a separate Isolate to be accessed via IsolateNameServer](https://drift.simonbinder.eu/docs/advanced-features/isolates/#independent-isolates):
```dart
Future<DriftIsolate> createIsolate() async {
  final token = RootIsolateToken.instance;
  return await DriftIsolate.spawn(() {
    BackgroundIsolateBinaryMessenger.ensureInitialized(token!);
    return LazyDatabase(() async {
      final dbFolder = await getApplicationDocumentsDirectory();
      final path = p.join(dbFolder.path, 'app_database.db');

      return NativeDatabase(File(path), logStatements: true);
    });
  });
}

//...

@DriftDatabase(tables: [SomeTable])
class AppDatabase extends _$AppDatabase {
  AppDatabase() : super(_openConnection());

  @override
  int get schemaVersion => 1;
}

LazyDatabase _openConnection() {
  driftRuntimeOptions.dontWarnAboutMultipleDatabases = true; // different queryexecutor used for each database
  return LazyDatabase(() async {
      final di = DriftIsolate.fromConnectPort(
      IsolateNameServer.lookupPortByName('drift_isolate')!);
    final conn = await di.connect();
    return conn.executor;
  });
}


//... main
  var di = await createIsolate();
  IsolateNameServer.registerPortWithName(di.connectPort, "drift_isolate");

//... some widget
  late AppDatabase db;
  late Stream<List<SomeTableData>> someStream;
  
  @override
  void initState() {
    super.initState();
    db = AppDatabase();
    someStream = db.select(db.someTable).watch();
  }
  
  @override
  void dispose() {
    super.dispose();
    db.close();
  }
  
  //... build
    StreamBuilder(
      stream: someStream,
      builder: ...,
    )
    
    //... some code in a different isolate with a different instance of AppDatabase
    await db.into(db.someTable).insert(...);
```

The initial `SELECT` from the `watch` function is logged, as is the `INSERT`, but the stream is never updated with the inserted entry. I must be misusing the API somehow. If anyone could point me in the right direction, it would be much appreciated.

**Comment by kuhnroyal** [comment]
If everything is correctly setup, you don't need `driftRuntimeOptions.dontWarnAboutMultipleDatabases = true;`. Are you calling `createIsolate()` somewhere else? You should only have one open database instance, and that is the single `DriftIsolate`. 
Your code looks okish but the problem may be in your insert code. Don't create a second isolate, connect with the same port from `IsolateNameServer.lookupPortByName('drift_isolate')`.

**Comment by cn-cn-cn** [comment]
> If everything is correctly setup, you don't need driftRuntimeOptions.dontWarnAboutMultipleDatabases = true;. Are you calling createIsolate() somewhere else?

No, I am only calling it once in main. I am opening multiple databases each connecting to the `DriftIsolate` because I need to use it from different isolates, as shown in the code.

> You should only have one open database instance, and that is the single `DriftIsolate`.

Yes, as far as I am aware, there is only one `DriftIsolate` created, and that is in main.

> Don't create a second isolate, connect with the same port from `IsolateNameServer.lookupPortByName('drift_isolate')`.

I am not, the `AppDatabase` creation code is as shown, it connects to the `DriftIsolate`. I create an `AppDatabase` when I need it and `close` it properly on disposal.

**Reply by kuhnroyal** [reply]
Like I said, if you see the warning and suppress it, you are likely doing something wrong.
You can have as many `AppDatabase` instances as you want, as long as they all connect to the same `DriftIsolate`, without the warning/suppression.

**Reply by kuhnroyal** [reply]
`dontWarnAboutMultipleDatabases` is about opening the underlying sqlite database multiple times. This happens when you have multiple `DriftIsolate`s but not when multiple `AppDatabase`s connect to the same `DriftIsolate`.

**Reply by cn-cn-cn** [reply]
> you are likely doing something wrong.

I realize that, but I am trying to determine what.

> You can have as many `AppDatabase` instances as you want, as long as they all connect to the same `DriftIsolate`, without the warning/suppression.

That's interesting, because the warning message claims otherwise, and is emitted by the `AppDatabase` instantiation, not from anything else:

```
I/flutter (13472): WARNING (drift): It looks like you've created the database class AppDatabase multiple times. When these two databases use the same QueryExecutor, race conditions will occur and might corrupt the database. 
I/flutter (13472): Try to follow the advice at https://drift.simonbinder.eu/faq/#using-the-database or, if you know what you're doing, set driftRuntimeOptions.dontWarnAboutMultipleDatabases = true
I/flutter (13472): Here is the stacktrace from when the database was opened a second time:
I/flutter (13472): #0      GeneratedDatabase._handleInstantiated (package:drift/src/runtime/api/db_base.dart:97:30)
I/flutter (13472): #1      GeneratedDatabase._whenConstructed (package:drift/src/runtime/api/db_base.dart:74:12)
I/flutter (13472): #2      new GeneratedDatabase (package:drift/src/runtime/api/db_base.dart:64:5)
(...)
I/flutter (13472): This warning will only appear on debug builds.
```

I am getting the warning despite the fact that I use a different `QueryExecutor` connecting to the `DriftIsolate` exposed through the `IsolateNameServer`.

**Reply by cn-cn-cn** [reply]
@kuhnroyal , could you tell me where the additional `DriftIsolate`s are being created? As far as I can see, only one DriftIsolate is created in

```dart
var di = await createIsolate();
IsolateNameServer.registerPortWithName(di.connectPort, "drift_isolate");
```

 which is then connected to in
 
 ```dart
final di = DriftIsolate.fromConnectPort(
IsolateNameServer.lookupPortByName('drift_isolate')!);
final conn = await di.connect();
 ```
 
 ? Or is my understanding of the code incorrect?

**Reply by kuhnroyal** [reply]
> That's interesting, because the warning message claims otherwise, and is emitted by the AppDatabase instantiation, not from anything else

Yea, you are right.
But multiple instances in different isolates (`main`, `DriftIsolate`, and you other isolates) don't share the database count and should still not emit this warning.

**Reply by kuhnroyal** [reply]
> Or is my understanding of the code incorrect?

No, this looks right. I am using it in roughly the same way.

**Reply by cn-cn-cn** [reply]
Right. I am creating another AppDatabase in the same Isolate. I will move that to one AppDatabase.

**Reply by kuhnroyal** [reply]
Are you using hot reload/restart?
Try to set `.connect(isolateDebugLog: true)`.

**Reply by cn-cn-cn** [reply]
No, I'm rebuilding every time. Will do, thanks.

**Reply by cn-cn-cn** [reply, marked as answer]
Thank you, using one AppDatabase per isolate resolved it For Some Reason™.
